### ISSUE TITLE
Explore MUSL to avoid GLIBC requirement

### ISSUE THREAD

**Issue by sunchao** [apache/arrow-datafusion-comet#138]
### What is the problem the feature request solves?

Currently we require dynamically linked glibc on the host machine which isn't desirable since it requires certain version of glibc, otherwise the native Comet lib won't be able to load. In particular, the glibc version used to compile and build Comet must be compatible (i.e., older) than the version used to execute Comet.

Another option is to explore `x86_64-unknown-linux-musl` compile target, which statically links against the [MUSL libc](https://musl.libc.org/) and can effectively remove the above constraints. However, t[here](https://andygrove.io/2020/05/why-musl-extremely-slow/) are some reports that MUSL can cause performance degradation, see [here](https://www.reddit.com/r/rust/comments/oh2k8l/rust_musl_and_glibc_in_2021/) and here. We may need to do some benchmark to evaluate.

### Describe the potential solution

Evaluate MUSL and check if it can be a potential candidate to co-exist (or replace) with the current GLIBC approach.

**Comment by snmvaughan** [comment]
I tend to agree with using a lower target glibc for compatibility.

If we want to consider MUSL we should also plan on benchmarking different combinations of target (glibc vs musl) and alloc (jemalloc vs system alloc) using a benchmark like TPC-DS.

**Comment by sunchao** [comment]
Yes, this ticket belongs to the "good to have" category IMO. One disadvantage of having to stick to the lowest common denominator for GLIBC is that certain crates may not able to compile. For instance, we ran into issues with `mimalloc` when building on GLIBC 2.17 IIRC.

**Comment by andygrove** [comment]
This recent blog post is worth a read. It talks about some performance issues with musl's default allocator and how switching to jemalloc helps.

https://blog.sdf.com/p/fast-development-in-rust-part-one

**Comment by viirya** [comment]
I remember we used jemalloc at the beginning, but we encountered some libraries issues from it and switched it back.
